crypto/cmp/cmp_ctx.c

```diff
@@ -0,0 +1,1143 @@
+}
+
+/*
+ * Set a callback function for log messages.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set_log_cb(OSSL_CMP_CTX *ctx, OSSL_cmp_log_cb_t cb)
+{
+    int res = 0;
+
+    if (ctx == NULL)
+        return 0;
+    if (cb == NULL) {
+#ifndef OPENSSL_NO_STDIO
+        BIO *bio_out = BIO_new_fp(stdout, BIO_NOCLOSE);
+
+        res = bio_out != NULL
+                  && OSSL_trace_set_channel(OSSL_TRACE_CATEGORY_CMP, bio_out);
+#endif
+    }
+    else {
+        res = OSSL_trace_set_callback(OSSL_TRACE_CATEGORY_CMP,
+                                      CMP_log_trace_cb, ctx);
+    }
+    if (res)
+        ctx->log_cb = cb;
+    return res;
+}
+
+/* print OpenSSL and CMP errors via the log cb of the ctx or OSSL_CMP_puts */
+void OSSL_CMP_print_errors(OSSL_CMP_CTX *ctx)
+{
+    OSSL_CMP_print_errors_cb(ctx == NULL ? NULL : ctx->log_cb);
+}
+
+/*
+ * Set or clear the reference value to be used for identification (i.e. the
+ * user name) when using PBMAC.
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_referenceValue(OSSL_CMP_CTX *ctx,
+                                     const unsigned char *ref, int len)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_INVALID_ARGS);
+        return 0;
+    }
+    return CMP_ASN1_OCTET_STRING_set1_bytes(&ctx->referenceValue, ref, len);
+}
+
+/*
+ * Set or clear the password to be used for protecting messages with PBMAC
+ * returns 1 on success, 0 on error
+ */
+int OSSL_CMP_CTX_set1_secretValue(OSSL_CMP_CTX *ctx, const unsigned char *sec,
+                                  const int len)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if (ctx->secretValue != NULL)
+        OPENSSL_cleanse(ctx->secretValue->data, ctx->secretValue->length);
+    return CMP_ASN1_OCTET_STRING_set1_bytes(&ctx->secretValue, sec, len);
+}
+
+/*
+ * Returns the stack of certificates received in a response message.
+ * The stack is duplicated so the caller must handle freeing it!
+ * returns pointer to created stack on success, NULL on error
+ */
+STACK_OF(X509) *OSSL_CMP_CTX_get1_extraCertsIn(const OSSL_CMP_CTX *ctx)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_INVALID_ARGS);
+        return NULL;
+    }
+    if (ctx->extraCertsIn == NULL)
+        return sk_X509_new_null();
+    return X509_chain_up_ref(ctx->extraCertsIn);
+}
+
+/*
+ * Copies any given stack of inbound X509 certificates to extraCertsIn of
+ * the OSSL_CMP_CTX structure so that they may be retrieved later.
+ * returns 1 on success, 0 on error
+ */
+int CMP_CTX_set1_extraCertsIn(OSSL_CMP_CTX *ctx,
+                              STACK_OF(X509) *extraCertsIn)
+{
+    if (ctx == NULL) {
+        CMPerr(0, CMP_R_NULL_ARGUMENT);
+        return 0;
+    }
+
+    sk_X509_pop_free(ctx->extraCertsIn, X509_free);
+    ctx->extraCertsIn = NULL;
+    if (extraCertsIn == NULL)
+        return 1;
+    return (ctx->extraCertsIn = X509_chain_up_ref(extraCertsIn)) != NULL;
```

Comment:
FIxed

Reply:
No, @Akretsch, we deliberately did not change our code here.

@mattcaswell, in my opinion each OpenSSL function should care to throw an error via the error queue whenever itself detects an error condition, rather than passing this as an undue burden to all callers, which - besides the multiplied effort - are typically not able to produce a to-the-point error message because they can only speculate what the actual error reason was.
So in calls like the above we as the caller should rely on the called function to behave properly.

I'm glad that functions like *`_new()` and *`_dup()` meanwhile do throw appropriate errors, e.g., an `ERR_R_MALLOC_FAILURE` on out of memory. Unfortunately this is not the case for (most of?) the *`_up_ref()` functions, and I'd say the responsibility for fixing this is with them.

So we when we call some non-CMP function we expect them to throw errors where needed, and we've made sure now that whenever our own functions in `cmp_ctx.c` and `cmp_util.c` detect an error they consistently call `CMPerr()`. Okay this way now?

Reply:
Ok - seems reasonable.